test/MinimalDelegation.execute.invariant.t.sol

```diff
@@ -0,0 +1,245 @@
+    }
+
+    /// @notice Executes a call with operation data (with signature)
+    /// @dev Handler function meant to be called during invariant tests
+    /// TODO: only supports single call arrays for now
+    /// - If the signing key is not registered on the account, expect the call to revert
+    function executeWithOpData(uint192 nonceKey, uint256 seed) public useSigningKey(seed) {
+        bool isRootKey = vm.addr(currentSigningKey.privateKey) == address(signerAccount);
+
+        bytes32 currentKeyHash = currentSigningKey.toKeyHash();
+        bool signatureIsValid;
+        if (!isRootKey) {
+            // TODO: check expiry here, settings, etc.
+            try signerAccount.getKey(currentKeyHash) {}
+            catch (bytes memory revertData) {
+                assertEq(bytes4(revertData), IKeyManagement.KeyDoesNotExist.selector);
+                signatureIsValid = false;
+            }
+        } else {
+            signatureIsValid = true;
+        }
+
+        HandlerCall memory handlerCall = _generateHandlerCall(seed);
+        HandlerCall[] memory handlerCalls = CallUtils.initHandler().push(handlerCall);
+
+        (uint256 nonce,) = _buildNextValidNonce(nonceKey);
+
+        Call[] memory calls = handlerCalls.toCalls();
+
+        bytes32 digest = signerAccount.hashTypedData(calls.toSignedCalls(nonce).hash());
+        bytes memory wrappedSignature =
+            abi.encode(isRootKey ? bytes32(0) : currentKeyHash, currentSigningKey.sign(digest));
+        bytes memory opData = abi.encode(nonce, wrappedSignature);
+        bytes memory executionData = abi.encode(calls, opData);
+
+        try signerAccount.execute(BATCHED_CALL_SUPPORTS_OPDATA, executionData) {
+            _processCallbacks(handlerCalls);
+        } catch (bytes memory revertData) {
+            if (!signatureIsValid) {
+                assertEq(bytes4(revertData), IKeyManagement.KeyDoesNotExist.selector);
+            } else if (handlerCall.revertData.length > 0) {
+                assertEq(revertData, handlerCall.revertData);
+            } else {
+                bytes memory debugCalldata =
+                    abi.encodeWithSelector(IERC7821.execute.selector, BATCHED_CALL_SUPPORTS_OPDATA, executionData);
+                console2.logBytes(debugCalldata);
+                console2.logBytes(revertData);
+                revert("uncaught revert");
+            }
+        }
+    }
+}
+
+/// @title MinimalDelegationExecuteInvariantTest
+contract MinimalDelegationExecuteInvariantTest is TokenHandler, DelegationHandler {
+    using KeyLib for Key;
+    using TestKeyManager for TestKey;
+    using CallUtils for Call;
+    using CallUtils for Call[];
+    using WrappedDataHash for bytes32;
+
+    MinimalDelegationExecuteInvariantHandler internal invariantHandler;
+
+    address public untrustedCaller = makeAddr("untrustedCaller");
+    uint256 public untrustedPrivateKey = 0xdead;
+
+    address internal sender = makeAddr("sender");
+
+    TestKey[] internal _keys;
```

Comment:
interesting so the _keys array here is really just used to pull a signer, but it seems like it would never test the state where we have created a new key (in _generate_generateHandlerCall) and then use that newly registered key to sign execution

Reply:
Yea, I can explore that in a follow up I think

Reply:
yeah I more so called it out cause I was confused by the different key states.. there is like fixture keys which I believe is just random keys that might get added/removed to the account and then keys which is just initialized with 2 to begin with (one root, one unauthorized) and those would actually sign new information

it might be helpful to separate them out by like signingKeys toRegisterKeys or something? and keys that are registered should be added to the signingKeys imo